### PR TITLE
Use path_for_whitelisting instead of canonical_path when matching

### LINKED ISSUE #1891: *_enum() takes item name after parse_callbacks processing instead of before
### Input C/C++ Header

```C++
enum Test {
    OptionA,
    OptionB,
    OptionC
};
```

### Bindgen Invocation
```Rust
use bindgen;
use std::env;
use std::path::PathBuf;

#[derive(Debug)]
struct RenameCallbacks {}

impl bindgen::callbacks::ParseCallbacks for RenameCallbacks {
    fn item_name(&self, original_variant_name: &str) -> Option<String> {
        if original_variant_name == "Test" {
            Some("AnotherName".into())
        } else {
            None
        }
    }
}

fn main() {
    let out_path = PathBuf::from(env::var("OUT_DIR").unwrap());
    let bindings = bindgen::Builder::default()
        .header("test.hpp")
        .parse_callbacks(Box::new(RenameCallbacks {}))
        .whitelist_type("Test")
        .bitfield_enum("Test")
        .generate()
        .expect("Couldn't generate bindings!");
    bindings
        .write_to_file(out_path.join("bindings.rs"))
        .expect("Couldn't write bindings!");
}
```

### Actual Results
```rust
pub const AnotherName_OptionA: AnotherName = 0;
pub const AnotherName_OptionB: AnotherName = 1;
pub const AnotherName_OptionC: AnotherName = 2;
pub type AnotherName = ::std::os::raw::c_int;
```

### Expected Results
Should generate bitfield enum. I figured out that changing the invocation to
```rust
    let bindings = bindgen::Builder::default()
        .header("test.hpp")
        .parse_callbacks(Box::new(RenameCallbacks {}))
        .whitelist_type("Test")
        .bitfield_enum("AnotherName")
        .generate()
        .expect("Couldn't generate bindings!");
```
will generate expected bindings.

This may not be a bug but it is definitely not documented. And it seems inconsistent with other APIs, since other APIs like `whitelist_type()` asks for name before `parse_callbacks` processing.

### PR DESCRIPTION
**Pull request by johnmave126**
for enum type specification and template opaque type specification.

Resolves #1891

### CODE REVIEW

**Comment by johnmave126** [comment]
Did look into having a test case, but seems unfit to current testing framework since there is no parse callback in command line.

**Review by emilio — APPROVED** [review]
This looks great, but we should be able to add a test to `bindgen-integration/`, which allows parse callbacks. If that's too hard for some reason though, please explain and I can merge as is.

Thank you!

**Comment by johnmave126** [comment]
Made two tests. One for enum and one for template.

File: bindgen-integration/build.rs [hunk 1]
```diff
@@ -50,9 +50,9 @@ impl ParseCallbacks for MacroCallback {
                 assert_eq!(value, b"string");
                 *self.seen_hellos.lock().unwrap() += 1;
             }
-            "TESTMACRO_STRING_EXPANDED" |
-            "TESTMACRO_STRING" |
-            "TESTMACRO_INTEGER" => {
+            "TESTMACRO_STRING_EXPANDED"
+            | "TESTMACRO_STRING"
+            | "TESTMACRO_INTEGER" => {
```
Comment (emilio) [inline bindgen-integration/build.rs:55]: This change seems unnecessary.

**Review by emilio — APPROVED** [review]
Sweet, thank you! I'll merge this manually to avoid that formatting change, but looks great otherwise.

**Comment by emilio** [comment]
f4d10c360453e91973cd08189af6185bb7c964c9

File: bindgen-integration/build.rs (hunk 1 above)
Comment (johnmave126) [reply bindgen-integration/build.rs:55]: Right. I bet rustfmt did this.

**Pull request closed without merging**